types/msg.go

```diff
@@ -77,8 +79,41 @@ type BankMsg struct {
 // SendMsg contains instructions for a Cosmos-SDK/SendMsg
 // It has a fixed interface here and should be converted into the proper SDK format before dispatching
 type SendMsg struct {
-	ToAddress   string `json:"to_address"`
-	Amount      Coins  `json:"amount"`
+	ToAddress string `json:"to_address"`
+	Amount    Coins  `json:"amount"`
+}
+
+type IbcMsg struct {
+	Ics20Transfer *Ics20TransferMsg `json:"ics20_transfer,omitempty"`
+	SendPacket    *SendPacketMsg    `json:"send_packet,omitempty"`
+	CloseChannel  *CloseChannelMsg  `json:"close_channel,omitempty"`
+}
+
+type Ics20TransferMsg struct {
+	ChannelID string `json:"channel_id"`
+	ToAddress string `json:"to_address"`
+	Amount    Coins  `json:"amount"`
+}
+
+type SendPacketMsg struct {
+	ChannelID        string           `json:"channel_id"`
+	Data             []byte           `json:"data"`
+	TimeoutHeight    IbcTimeoutHeight `json:"timeout_timestamp"`
```

Comment:
Req: Should be a pointer as it is either height or timestamp for the timeout

Reply:
Req: json tag should not be `timeout_timestamp`

Reply:
Okay, so Timeout height is (version, height). Timeout timestamp is in seconds (millisecond? microseconds?) since UNIX 0 ??? Maybe I use `time.Time` in Go and custom marshal/unmarshal logic?

And only one of the two can be set per packet? Is exactly one required or is leaving both blank also valid?

Reply:
One of them is mandatory https://github.com/cosmos/cosmos-sdk/blob/v0.40.0/x/ibc/core/04-channel/types/packet.go#L105
Btw. the `TimeoutHeight.IsZero()` looks very helpful. ;-)

Reply:
Ah, good eye. So, 1 or 2 are required. Not a case for an enum, but a clear `IsZero`